packages/tailwindcss/src/utilities.ts

```diff
@@ -1261,19 +1275,47 @@ export function createUtilities(theme: Theme) {
     ],
   })
 
+  let rotateProperties = () =>
+    atRoot([
+      property('--tw-rotate', '0deg', '<angle>'),
+      // Vector components for the axis of rotation
+      property('--tw-rotate-x', '0', '<number>'),
+      property('--tw-rotate-y', '0', '<number>'),
+      property('--tw-rotate-z', '0', '<number>'),
```

Comment:
In my head I was thinking we'd just have two variables:

```
property('--tw-rotate', '0deg', '<angle>'),
property('--tw-rotate-axis'),
```

Then the values for `--tw-rotate-axis` would be something like:

`[x | y | z | <number{3}]`

...however you do that with `syntax` in `@property`.

This means you couldn't do this:

```
<div class="rotate-45 rotate-x rotate-y">
```

...but I think that doesn't bother me personally, you'd just have to use a vector for that:

```
<div class="rotate-45 rotate-[1_1_0]">
```

I'm not sure how common it actually is to want to rotate something by exactly the same angle on two different axis — my gut is probably not common?

With this change we wouldn't have to translate `rotate-x` to a vector, it could just set `--tw-rotate-axis: x` because `rotate: x 45deg` is valid.

Reply:
To add to this — we could (later) add `rotate-xy` as a utility too. Which seems just as explanatory as two separate utilities. So I'm Adam on this one. A single property for the axis feels like a good way to do this.

Reply:
Would we still be able to do something like `rotate-x-45 rotate-y-135 rotate-z-20`?

Having granular control over individual axis transforms in a pragmatic way rather than using a vector is an important feature here IMO.

This is possible, and can even be converted into a vector under the hood if need be, which I demo'd here:
https://x.com/branmcconnell/status/1768009776938844550

Reply:
We've further simplified this by specifying the rotation axis via a modifier on `rotate`:

| Class | CSS |
| ---- | ---- |
| `rotate-45` | `rotate: 45deg` |
| `rotate-45/x` | `rotate: x 45deg` |
| `rotate-45/y` | `rotate: y 45deg` |
| `rotate-45/z` | `rotate: z 45deg` |
| `rotate-45/[1_2_3]` | `rotate: 1 2 3 45deg` |

This makes it clear angles are not composable, in line with what CSS supports via the [`rotate` property](https://developer.mozilla.org/en-US/docs/Web/CSS/rotate).

Composing multiple angles requires a pipeline using `transform`. That is now supported via arbitrary `transform` values (e.g. `transform-[rotateX(45deg) rotateY(-90deg)]`).

Reply:
@KrisBraun Yes, I was primarily speaking about using multi-axis rotation using the traditional syntax instead of the vector syntax.

This is possible using even the regular `rotate` properties if you convert them to vector syntax under the hood. I demonstrate this in the twitter thread I linked above ([also here](https://x.com/branmcconnell/status/1768009776938844550)).

So hypothetically, someone could do this and get the expected behavior:
```
rotate-45/x rotate-30/y rotate-60/z
```

Reply:
After more thought, we decided `rotate-x` and `rotate-y` are more in line with other Tailwind utilities, even if they are mutually exclusive. The PR and description have been updated.

Reply:
@brandonmcconnell Totally understand the appeal of composable rotation angles. They're achievable using an arbitrary `transform` value as shown above. It's more verbose but has at least a couple advantages:

1. The order of rotations is explicitly specified.
2. It's clear the `transform` property is being modified, rather than `rotate`.

There are also a few reasons not to translate multiple rotations into a vector:

1. CSS variables can't be supported if the translation is done in code at compile time.
2. Implementing the translation with CSS math functions is technically possible, but ridiculously long and likely slow.
3. Any translation would need to assume an order of composition, unless that was also configurable, ballooning the complexity.
4. The translated values would be harder to understand in dev tools.

We're going with the grain of CSS on this. The `rotate` property is clearly designed for applying a single angle of rotation, and `transform` is the escape hatch for more complex transformation pipelines.

Thanks so much for initiating this PR! We're now at a point where the simple things are simple and everything else is possible. We're looking forward to seeing what folks build with this!